waku/node/peer_manager/peer_manager.nim

```diff
@@ -619,30 +606,58 @@ proc getNumStreams*(pm: PeerManager, protocol: string): (int, int) =
   return (numStreamsIn, numStreamsOut)
 
 proc pruneInRelayConns(pm: PeerManager, amount: int) {.async.} =
-  let (inRelayPeers, outRelayPeers) = pm.connectedPeers(WakuRelayCodec)
+  let (inRelayPeers, _) = pm.connectedPeers(WakuRelayCodec)
   let connsToPrune = min(amount, inRelayPeers.len)
 
   for p in inRelayPeers[0..<connsToPrune]:
     asyncSpawn(pm.switch.disconnect(p))
 
 proc connectToRelayPeers*(pm: PeerManager) {.async.} =
-  let (inRelayPeers, outRelayPeers) = pm.connectedPeers(WakuRelayCodec)
-  let maxConnections = pm.switch.connManager.inSema.size
-  let totalRelayPeers = inRelayPeers.len + outRelayPeers.len
-  let inPeersTarget = maxConnections - pm.outRelayPeersTarget
+  var peersToConnect: seq[RemotePeerInfo]
+  var peersToDisconnect: int
+
+  for shard in pm.wakuMetadata.shards.items:
+    let peers = pm.peerStore.getPeersByShard(uint16(pm.wakuMetadata.clusterId), uint16(shard))
+
+    let connectedInPeers = peers.filterIt(
+      pm.peerStore.isConnected(it.peerId) and it.direction == Inbound)
+    
+    let connectedOutPeers = peers.filterIt(
+      pm.peerStore.isConnected(it.peerId) and it.direction == Outbound)
+
+    let outPeerDiff = pm.outRelayPeersTarget - connectedOutPeers.len
+    let inPeerDiff = pm.inRelayPeersTarget - connectedInPeers.len
 
-  # TODO: Temporally disabled. Might be causing connection issues
-  #if inRelayPeers.len > pm.inRelayPeersTarget:
-  #  await pm.pruneInRelayConns(inRelayPeers.len - pm.inRelayPeersTarget)
+    if inPeerDiff <= 0:
+      peersToDisconnect += abs(inPeerDiff)
 
-  if outRelayPeers.len >= pm.outRelayPeersTarget:
+    if outPeerDiff <= 0:
+      continue
+
+    let connectablePeers = peers.filterIt(
+      (not pm.peerStore.isConnected(it.peerId) and pm.canBeConnected(it.peerId)))
+
+    peersToConnect.insert(connectablePeers[0..outPeerDiff])
```

Comment:
If peers are inserted to the peersToConnect seq by shard based on the order in `pm.wakuMetadata.shards.items` seq, if one of the first shards on that array always have a lot of peers to connect to, will other shards get starved? If so, is that an issue?

Reply:
Each shard should be independent, and with an in/out target # of peers. If some shard has less peers to connect to it's a discovery problem, not management.

To answer your question, yes they could but can't fix that since discovery is random.

Reply:
Just notice that `connectablePeer` could be shorter than `outPeerDif`

Reply:
Makes sense, thank you!

> Just notice that `connectablePeer` could be shorter than `outPeerDif`

True! Great catch! :)